HDDS-11925. Fix deploy step in Release Manager guide

adoroszlai opened the pull request:
## What changes were proposed in this pull request?

`mvn deploy` rebuilds the project, as it executes all [phases of the default lifecycle](https://maven.apache.org/ref/3.9.9/maven-core/lifecycles.html#default-lifecycle) up to `deploy`.  Not only is it unnecessary, it causes jars to be [different between Maven repo and binary tarball](https://github.com/apache/ozone/discussions/7560).

`mvn deploy:deploy` just executes Maven Deploy Plugin's [`deploy` task](https://maven.apache.org/plugins/maven-deploy-plugin/usage.html#the-deploy-deploy-mojo), which is to upload artifacts to the remote repository.

Flags `-Psign -DskipTests -Dbuildhelper.skipAttach` can be removed, since they are specific to other plugins, which would (before this change) be executed in various lifecycle phases.

Skipping `:ozone-dist` seems to be unnecessary, too.  I don't see why we should not include it in the Maven repo.

**Update**: the proposed solution described above does not work, please see [this comment](https://github.com/apache/ozone-site/pull/109#issuecomment-2572923357) for current one.

https://issues.apache.org/jira/browse/HDDS-11925

## How was this patch tested?

Will be tested during next release.  (Tested similar change in Ratis, see https://github.com/apache/ratis/pull/1188.)

errose28 commented:
Thanks for looking into this @adoroszlai. I'm also not sure why `ozone-dist` was skipped originally. Maybe this was an attempt to re-use those artifacts from the [original build](https://ozone-site-v2.staged.apache.org/docs/developer-guide/project/release-guide#create-the-release-artifacts)?

Is there an extra verification step we can add to the release guide to make sure something like this gets caught during the release cycle? For example if some part of the original build is accidentally cleaned and the `mvn deploy:deploy` uploads partial artifacts. Maybe after [uploading to nexus](https://ozone-site-v2.staged.apache.org/docs/developer-guide/project/release-guide#upload-the-artifacts-to-apache-nexus) we can have the manager download the artifacts from both sources and check if they match.

adoroszlai commented:
Updated the PR:

- Steps now include two separate builds, same as previously:
   1. `mvn clean install` to fetch dependencies and ensure that the build is successful
   2. `mvn deploy`, like originally, to produce the artifacts that get published.  Added `-DdeployAtEnd=true` to avoid partial build getting deployed.
- The key difference is that tarballs are copied from the second build, not the first one.  This ensures jars in the tarball are the same as those published to Maven.
- Moved initial setup (Maven configuration) to Prerequisites.
- Also moved comment about GNU vs. BSD tar to Prerequisites.
- Split pre-build `rat` check to new section "Sanity Checks", which can be expanded later if needed.

kerneltime commented:
Offtopic: Why can't this be automated by pushing a tag against a branch to build and publish the artifacts using GitHub actions?

adoroszlai commented:
> Why can't this be automated

We can and should automate the release process as script(s) (e.g. Ratis has more automation than Ozone), but

> by pushing a tag against a branch to build and publish the artifacts using GitHub actions?

we cannot publish fully by automation, as per [Apache policy](https://infra.apache.org/release-publishing.html#signing).

jojochuang reviewed (approved):
+1 confirmed the jars uploaded to apache maven repo is the same as the ones in the tarball.

adoroszlai commented:
Thanks @errose28, @jojochuang for the review.
